### PR TITLE
Better handle hyperlinks and colors

### PR DESCRIPTION
**Pull request by romainfrancois**
### Intent

addresses #10977

### Approach

hyperlinks now only get the `xtermHyperlink` css class. 

### Automated Tests

> Indicate whether this change includes unit tests or integration tests, or link a work item or pull request to add those tests to another repo. If the change cannot or will not be covered by a test, indicate why.

### QA Notes

```r
pak::pak("r-lib/cli")

library(cli)

link <- cli::style_hyperlink("test", "http://rstudio.com")

# magenta because unset
link

# already set to red, so magenta not applied
cli::col_red(link)

# set to magenta, but then overruled by red
cli::style_hyperlink(cli::col_red("test"), "http://rstudio.com")
```

gets: 

image

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

**Comment by romainfrancois** [comment]
It is however currently only possible to cancel underline from inside: 

image

I guess I need to better understand how `style_no_underline()` works: 

```
> str(cli::style_no_underline)
function (...)  
 - attr(*, "class")= chr [1:2] "cli_ansi_style" "ansi_style"
 - attr(*, "_styles")=List of 1
  ..$ no_underline:List of 3
  .. ..$ open   : chr "\033[0m\033[22m\033[23m\033[27m\033[28m\033[29m\033[39m\033[49m"
  .. ..$ close  : chr "\033[24m"
  .. ..$ palette: NULL
```

**Comment by romainfrancois** [comment]
But are we even supposed to be able to cancel the underline ? i.e. in iTerm we get

image

Should we do the same ?

**Comment by gaborcsardi** [comment]
Yeah, I am surprised that cancelling the underline works at all. I don't think we need to worry about it too much. We'll not cancel it, anyway, we'll just got with the default style. (Which probably does not need to add an explicit underline.)

**Comment by romainfrancois** [comment]
Ok thanks. I'll simplify so that: 
  - it's always underline and governed by the `xtermHyperlink` style so that themes can decide differently. 
  - but colors can be set either way: `col_red(link)` or `style_hyperlink(col_red(text), url)`

File: src/gwt/src/org/rstudio/core/client/VirtualConsole.java [hunk 1]
```diff
@@ -29,7 +29,6 @@
 import org.rstudio.core.client.regex.Match;
 import org.rstudio.core.client.regex.Pattern;
 import org.rstudio.core.client.virtualscroller.VirtualScrollerManager;
-import org.rstudio.studio.client.RStudioGinjector;
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/core/client/VirtualConsole.java:32]: Can we add some unit tests for this scenario? The ANSI / Console stuff handling code is fragile and complex so having lots of coverage is a good idea. Lots of examples in: https://github.com/rstudio/rstudio/blob/main/src/gwt/test/org/rstudio/core/client/VirtualConsoleTests.java

**Review by gtritchie — CHANGES_REQUESTED** [review]

**Comment by hadley** [comment]
I like that styling 😄

File: src/gwt/src/org/rstudio/core/client/VirtualConsole.java [hunk 2]
```diff
@@ -609,19 +608,31 @@ public void submit(String data, String clazz, boolean forceNewRange, boolean ari
...
                   if (!StringUtil.equals(url, ""))
                   {
-                     ansiCodeStyles_ = ansi_.processCode("\033[4m"); // underline
-                     ansiCodeStyles_ = ansi_.processCode("\033[35m"); // magenta
-                     currentClazz = setCurrentClazz(ansiColorMode, clazz);
+                     hyperlink_ = new Hyperlink(url, /*params=*/ hyperlinkMatch.getGroup(1), !currentClazz.contains("xtermColor"), !currentClazz.contains("xtermUnderline"));
+
+                     // add magenta if the text is not already coloured
+                     if (hyperlink_.setColor)
+                        ansiCodeStyles_ = ansi_.processCode("\033[35m"); // magenta
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply src/gwt/src/org/rstudio/core/client/VirtualConsole.java]: This no longer needs to `processCode`

File: src/gwt/src/org/rstudio/core/client/VirtualConsole.java [hunk 3]
```diff
@@ -609,19 +608,31 @@ public void submit(String data, String clazz, boolean forceNewRange, boolean ari
...
                   {
-                     ansiCodeStyles_ = ansi_.processCode("\033[39m"); // </magenta>
-                     ansiCodeStyles_ = ansi_.processCode("\033[24m"); // </underline>
+                     // close underline only if it has been added by the hyperlink
+                     if (hyperlink_.setUnderline)
+                        ansiCodeStyles_ = ansi_.processCode("\033[24m"); // underline
+                     
+                     // reset colour only if it has been added by the hyperlink
+                     if (hyperlink_.setColor)
+                        ansiCodeStyles_ = ansi_.processCode("\033[39m"); // magenta
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply src/gwt/src/org/rstudio/core/client/VirtualConsole.java]: Thanks. Although this no longer uses it.

File: src/gwt/src/org/rstudio/core/client/VirtualConsole.java (hunk 1 above)
Comment (romainfrancois) [reply src/gwt/src/org/rstudio/core/client/VirtualConsole.java:32]: I added two, but I currently struggle with my java settings ...: 

```
romainfrancois@mezcalita gwt % ./ant unittest
Building with Java at /opt/homebrew/opt/openjdk@11
Using JAVA_HOME: /opt/homebrew/opt/openjdk@11
Buildfile: /Users/romainfrancois/git/rstudio/src/gwt/build.xml

build-unittests:

unittest:
     [java] JUnit version 4.9b3
     [java] Exception in thread "main" java.lang.UnsupportedClassVersionError: org/rstudio/studio/client/RStudioUnitTestSuite has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0
     [java] 	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
     [java] 	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
     [java] 	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
     [java] 	at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:800)
     [java] 	at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:698)
     [java] 	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:621)
     [java] 	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:579)
     [java] 	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
     [java] 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
     [java] 	at java.base/java.lang.Class.forName0(Native Method)
     [java] 	at java.base/java.lang.Class.forName(Class.java:315)
     [java] 	at org.junit.runner.JUnitCore.runMain(JUnitCore.java:89)
     [java] 	at org.junit.runner.JUnitCore.runMainAndExit(JUnitCore.java:53)
     [java] 	at org.junit.runner.JUnitCore.main(JUnitCore.java:45)

BUILD FAILED
/Users/romainfrancois/git/rstudio/src/gwt/build.xml:358: Java returned: 1

Total time: 0 seconds
```

Comment (kevinushey) [reply src/gwt/src/org/rstudio/core/client/VirtualConsole.java:32]: Do you have any luck with `./ant clean && ./ant unittest`? If not, we might need to tweak how JAVA_HOME is found and set in the `ant` helper script.

Comment (romainfrancois) [reply src/gwt/src/org/rstudio/core/client/VirtualConsole.java:32]: `./ant clean` typically helps for `./ant devmode` but for some reason not for `./ant unittest`. I'll investigate.

**Comment by romainfrancois** [comment]
yeah, getting: 

```
... 
     [java] Time: 12,743
     [java]
     [java] OK (213 tests)
     [java]

BUILD SUCCESSFUL
Total time: 13 seconds
```

**Pull request merged**